emma-language/src/test/scala/org/emmalanguage/compiler/lang/opt/FoldForestFusionSpec.scala

```diff
@@ -0,0 +1,543 @@
+          val res3 = rds3 fold alg$FlatMap$r1
+          val res1 = rds1 fold alg$Map$r1
+          val res2 = app$Min$r2.get
+          val res4 = (res1, res2, res3)
+          res4
+        })
+
+        testPipeline(Core.anf)(src) shouldBe alphaEqTo(exp)
+      }
+
+      "in Emma Core" in {
+        val exp = anfPipeline(reify {
+          val `+/- 5$r3` : this.`+/- 5`.type = this.`+/- 5`
+          val `_ > 0$r4` : this.`_ > 0`.type = this.`_ > 0`
+          val alg$Min$r3 = alg.Min(scala.math.Ordering.Int)
+          val alg$Sum$r3 = alg.Sum(scala.math.Numeric.IntIsIntegral)
+          val `is prime$r3` : this.`is prime`.type = this.`is prime`
+          val rds1 = this.rands
+          val rds2 = this.rands
+          val rds3 = this.rands
+          val square$r3: this.square.type = this.square
+          val alg$Count$r3 = alg.Count(`is prime$r3`)
+          val alg$WithFilter$r2 = alg.WithFilter(`_ > 0$r4`, alg$Min$r3)
+          val fun$FlatMap$r1 = (x$r6: scala.Int) => {
+            val xs$r1 = comprehension[scala.Int, DataBag]({
+              val x$r12 = generator[scala.Int, DataBag]({
+                val x$r13 = `+/- 5$r3`(x$r6)
+                x$r13
+              })
+              head[scala.Int]({
+                x$r12
+              })
+            })
+            xs$r1
+          }
+          val alg$FlatMap$r2 = alg.FlatMap(fun$FlatMap$r1, alg$Sum$r3)
+          val alg$Map$r2 = alg.Map(square$r3, alg$Count$r3)
+          val app$Min$r3 = rds2.fold(alg$WithFilter$r2)
+          val res1 = rds1.fold(alg$Map$r2)
+          val res2 = app$Min$r3.get
+          val res3 = rds3.fold(alg$FlatMap$r2)
+          val res4 = (res1, res2, res3)
+          res4
+        })
+
+        testPipeline(Core.lift)(src) shouldBe alphaEqTo(exp)
+      }
+    }
+
+    "of depth two" - {
+      val src = u.reify {
+        rands.withFilter(`_ > 0`).map(`x => x * x`).sum
+      }.tree
+
+      "in Emma LNF" in {
+        val exp = anfPipeline(reify {
+          val `_ > 0$r3`: this.`_ > 0`.type = this.`_ > 0`
+          val alg$Sum$r3 = alg.Sum(Numeric.IntIsIntegral)
+          val rands$r5: this.rands.type = this.rands
+          val `x => x * x$r1`: this.`x => x * x`.type = this.`x => x * x`
+          val alg$Map$r2 = alg.Map(`x => x * x$r1`, alg$Sum$r3)
+          val alg$WithFilter$r2 = alg.WithFilter(`_ > 0$r3`, alg$Map$r2)
+          val sum$r3 = rands$r5 fold alg$WithFilter$r2
+          sum$r3
+        })
+
+        testPipeline(Core.anf)(src) shouldBe alphaEqTo(exp)
+      }
+
+      "in Emma Core" in {
+        val exp = anfPipeline(reify {
+          val `_ > 0$r1` : this.`_ > 0`.type = this.`_ > 0`
+          val alg$Sum$r1 = alg.Sum(Numeric.IntIsIntegral)
+          val rands$r1: this.rands.type = this.rands
+          val `x => x * x$r1` : this.`x => x * x`.type = this.`x => x * x`
+          val fun$FlatMap$r1 = (x$r1: scala.Int) => {
+            val xs$r1 = comprehension[scala.Int, DataBag]({
```

Comment:
This is not correct (missing generator), because we didn't think of the case `xs.withFilter(p).map(f)`.